core/src/main/java/org/apache/iceberg/BaseManifestReader.java

```diff
@@ -0,0 +1,294 @@
+ * @param <ThisT> The Java class of this reader, returned by configuration methods.
+ */
+abstract class BaseManifestReader<T, ThisT> extends CloseableGroup implements CloseableIterable<T> {
+  static final ImmutableList<String> ALL_COLUMNS = ImmutableList.of("*");
+  private static final Set<String> STATS_COLUMNS = Sets.newHashSet(
+      "value_counts", "null_value_counts", "lower_bounds", "upper_bounds");
+
+  protected enum FileType {
+    DATA_FILES(GenericDataFile.class.getName()),
+    DELETE_FILES("...");
+
+    private final String fileClass;
+
+    FileType(String fileClass) {
+      this.fileClass = fileClass;
+    }
+
+    private String fileClass() {
+      return fileClass;
+    }
+  }
+
+  private final InputFile file;
+  private final InheritableMetadata inheritableMetadata;
+  private final FileType content;
+  private final Map<String, String> metadata;
+  private final PartitionSpec spec;
+  private final Schema fileSchema;
+
+  // updated by configuration methods
+  private Expression partFilter = alwaysTrue();
+  private Expression rowFilter = alwaysTrue();
+  private Schema fileProjection = null;
+  private Collection<String> columns = null;
+  private boolean caseSensitive = true;
+
+  // lazily initialized
+  private Evaluator lazyEvaluator = null;
+  private InclusiveMetricsEvaluator lazyMetricsEvaluator = null;
+
+  protected BaseManifestReader(InputFile file, Map<Integer, PartitionSpec> specsById,
+                               InheritableMetadata inheritableMetadata, FileType content) {
+    this.file = file;
+    this.inheritableMetadata = inheritableMetadata;
+    this.content = content;
+
+    try {
+      try (AvroIterable<ManifestEntry> headerReader = Avro.read(file)
+          .project(ManifestEntry.getSchema(Types.StructType.of()).select("status"))
+          .build()) {
+        this.metadata = headerReader.getMetadata();
+      }
+    } catch (IOException e) {
+      throw new RuntimeIOException(e);
+    }
+
+    int specId = TableMetadata.INITIAL_SPEC_ID;
+    String specProperty = metadata.get("partition-spec-id");
+    if (specProperty != null) {
+      specId = Integer.parseInt(specProperty);
+    }
+
+    if (specsById != null) {
+      this.spec = specsById.get(specId);
+    } else {
+      Schema schema = SchemaParser.fromJson(metadata.get("schema"));
+      this.spec = PartitionSpecParser.fromJsonFields(schema, specId, metadata.get("partition-spec"));
+    }
+
+    this.fileSchema = new Schema(DataFile.getType(spec.partitionType()).fields());
+  }
+
+  protected abstract ThisT self();
+
+  public InputFile file() {
+    return file;
+  }
+
+  public Schema schema() {
+    return fileSchema;
+  }
+
+  public PartitionSpec spec() {
+    return spec;
+  }
+
+  public ThisT select(Collection<String> newColumns) {
+    Preconditions.checkState(fileProjection == null,
+        "Cannot select columns using both select(String...) and project(Schema)");
+    this.columns = newColumns;
+    return self();
```

Comment:
curious to know, why don't we simply return `this` in these situations?

Reply:
This needs to return the subclass, which it can't guarantee to be `ThisT`. The `self` method returns the correct type because it is overridden in each subclass to return `this`.